### PR TITLE
JDK-8255603: Memory/Performance regression after JDK-8210985

### PR DESCRIPTION
**Pull request by RealCLanger**
It seems that there exists a memory/performance regression that was introduced with JDK-8210985: Update the default SSL session cache size to 20480.

The idea to limit the maixmum SSL session cache size by itself is good. Unfortunately, as per the current implementation of sun.security.util.MemoryCache, it also modifies the initial size of the LinkedHashMap that is backing the cache to initialize with more than the maximum size.

I suggest to restore the original behavior that initializes with an initialCapacity of 1 in most cases. That was true when before JDK-8210985, the property javax.net.ssl.sessionCacheSize was not set at all.
In case it was set, the initial size would have been like now, (javax.net.ssl.sessionCacheSize / 0.75f) + 1, which still seems strange.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (4/4 passed) |

### Issue
 * [JDK-8255603](https://bugs.openjdk.java.net/browse/JDK-8255603): Memory/Performance regression after JDK-8210985

### Reviewers
 * [Volker Simonis](https://openjdk.java.net/census#simonis) (@simonis - **Reviewer**)
 * [Xue-Lei Andrew Fan](https://openjdk.java.net/census#xuelei) (@XueleiFan - **Reviewer**)
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/937/head:pull/937`
`$ git checkout pull/937`

### CODE REVIEW

**Comment by XueleiFan** [comment]
Did you have a benchmark with various cache sizes (for example, from 1 to 10K) and various connections (for example from 1 to 10K) for those components (including TLS implementation) that use Cache?

**Comment by RealCLanger** [comment]
> Did you have a benchmark with various cache sizes (for example, from 1 to 10K) and various connections (for example from 1 to 10K) for those components (including TLS implementation) that use Cache?

Nope, we've just seen the memory regression in a certain customer use case (lot's of meory retained by a finalizer) and confirmed it resolved after setting javax.net.ssl.sessionCacheSize to 0.

But I guess this change merits certain further benchmarking to get it right.

**Comment by simonis** [comment]
Congratulations! You beat me about half an hour with this issue :)

I was debugging this regression already the whole day and when I searched for JDK-8210985 in my mailbox I saw your post.

**Comment by XueleiFan** [comment]
> > Did you have a benchmark with various cache sizes (for example, from 1 to 10K) and various connections (for example from 1 to 10K) for those components (including TLS implementation) that use Cache?
> 
> Nope, we've just seen the memory regression in a certain customer use case (lot's of meory retained by a finalizer) and confirmed it resolved after setting javax.net.ssl.sessionCacheSize to 0.
> 
Did you have this patch checked with the customer?  I think the performance may be similar or improved comparing to set the cache size to 0.

> But I guess this change merits certain further benchmarking to get it right.
It looks good to me, but we may be more confident with it if there is a benchmarking.

**Comment by RealCLanger** [comment]
We're currently rolling out a patch to our SAP JVM shipment (based on Oracle's JDK 8 licensee repository) with exactly this content. We will then check with the customer but I'd suspect his results will about the same as with -Djavax.net.ssl.sessionCacheSize=0.

If you require some benchmarking I guess it'll take me some more time.

In the end I doubt that we'll find a better default value than 1 for the cache size as it's hard to predict how full a cache will be in the average. Maybe one could spend a property for the initial size - but I also doubt that's worth the effort.

I also think that for the use case in StatusResponseManager's responseCache the influence of a different initial value is neglectable.

**Comment by simonis** [comment]
I can confirm that JDK-8210985 can cause massive regressions in memory consumption. Also, [JDK-8253116
Performance regression observed post upgrade to 8u261](https://bugs.openjdk.java.net/browse/JDK-8253116) is clearly a duplicate for this.

I'm fine with setting the initial cache size to 1 as this restores the original behavior for `javax.net.ssl.sessionCacheSize=0`.

The other possibility would be to use the default size for `LinkedHashMap` but that's not easy if we want to set the `accessOrder` because the only constructor which takes the `accessOrder` also requires the specification of `initialCapacitiy` and `loadFactor`. Otherwise, `LinkedHashMap`s capacity defaults to `HashMap`s default capacity which is `16`.

Setting the `initialCapacity` to `1` will initialize the hash map with one bucket (see initialization code in `HashMap`).
```
    /**
     * Returns a power of two size for the given target capacity.
     */
    static final int tableSizeFor(int cap) {
        int n = cap - 1;
        n |= n >>> 1;
        n |= n >>> 2;
        n |= n >>> 4;
        n |= n >>> 8;
        n |= n >>> 16;
        return (n < 0) ? 1 : (n >= MAXIMUM_CAPACITY) ? MAXIMUM_CAPACITY : n + 1;
    }
```
Benchmarking is probably hard because we don't know the average occupancy of the map.

So in the end I think `1` is a good solution for now.

**Review by simonis — APPROVED** [review]

**Comment by XueleiFan** [comment]
> Benchmarking is probably hard because we don't know the average occupancy of the map.
I agreed. No matter what the default value is, it will not fit perfectly in all situations.  The value 1 may be fit for small workload applications, but not good for big workload applications.  Applications could use the size setting APIs for the tuning.  For this update, I think the impact for various workload may be limited/acceptable, but I'm not very sure of it.  Benchmarking data with various workload would help us for a better sense.

**Review by XueleiFan — CHANGES_REQUESTED** [review]
No regression test.

**Comment by simonis** [comment]
> > Benchmarking is probably hard because we don't know the average occupancy of the map.
> 
> I agreed. No matter what the default value is, it will not fit perfectly in all situations. The value 1 may be fit for small workload applications, but not good for big workload applications. Applications could use the size setting APIs for the tuning. For this update, I think the impact for various workload may be limited/acceptable, but I'm not very sure of it. Benchmarking data with various workload would help us for a better sense.

But we did run with `1` for quite a long time without somebody complaining :)

**Comment by XueleiFan** [comment]
> > > Benchmarking is probably hard because we don't know the average occupancy of the map.
> > 
> > 
> > I agreed. No matter what the default value is, it will not fit perfectly in all situations. The value 1 may be fit for small workload applications, but not good for big workload applications. Applications could use the size setting APIs for the tuning. For this update, I think the impact for various workload may be limited/acceptable, but I'm not very sure of it. Benchmarking data with various workload would help us for a better sense.
> 
> But we did run with `1` for quite a long time without somebody complaining :)

Yes, I think it is a safe update and looks good to me.  I believe the impact should be minimal.  But normally, I would like to check with a test for sure.  If no regression test, an explain with noreg tag may be needed.  External testing, like a confirmation of no performance regression any longer in an existing application,  is fine.

I don't want to block this integration, please go ahead if you are confident with it.

**Comment by RealCLanger** [comment]
I added the noreg-hard label to the JBS bug. I'll wait for any further input til monday, before integration.

@XueleiFan would be nice if you could approve then, too :)

**Review by XueleiFan — APPROVED** [review]
I had a benchmarking with one client and one server for full handshaking, with JMH.  I did not see throughput regression with this benchmarking, and the improvement is not visible although.  The benchmarking itself may be limited as it is trying to use a large volume of connections.

File: src/java.base/share/classes/sun/security/util/Cache.java [hunk 1]
```diff
@@ -265,8 +265,7 @@ public MemoryCache(boolean soft, int maxSize, int lifetime) {
         else
             this.queue = null;
 
-        int buckets = (int)(maxSize / LOAD_FACTOR) + 1;
-        cacheMap = new LinkedHashMap<>(buckets, LOAD_FACTOR, true);
+        cacheMap = new LinkedHashMap<>(1, LOAD_FACTOR, true);
```
Comment (theRealAph) [inline src/java.base/share/classes/sun/security/util/Cache.java:268]: This looks right. The idea of scaling the initial cache size to the maximum size seems obviously to be wrong.

**Review by theRealAph — APPROVED** [review]

**Pull request closed without merging**